cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java

```diff
@@ -1068,8 +1071,15 @@ private synchronized void stopAndRemoveWatchers(String namespace) {
   /**
    * Stop and remove watchers for all Kubernetes namespaces
    */
-  private synchronized void stopAndRemoveWatchers() {
-    resourceWatchers.keySet().forEach(this::stopAndRemoveWatchers);
+  @VisibleForTesting
+  synchronized void stopAndRemoveWatchers() {
+    Iterator<Entry<String, Map<Type, AppResourceWatcherThread<?>>>> iter =
+        resourceWatchers.entrySet().iterator();
+    while (iter.hasNext()) {
+      Map<Type, AppResourceWatcherThread<?>> watchers = iter.next().getValue();
+      watchers.values().forEach(AbstractWatcherThread::close);
+      iter.remove();
```

Comment:
iterators don't always implement the remove method. It's simpler to just clear the collection at the very end after iterating through.